tunnel/outline.go

```diff
@@ -16,65 +16,87 @@ package tunnel
-	"github.com/eycorsican/go-tun2socks/proxy/socks"
+
+	oss "github.com/Jigsaw-Code/outline-go-tun2socks/shadowsocks"
+	"github.com/Jigsaw-Code/outline-ss-server/shadowsocks"
 )
 
 // OutlineTunnel represents a tunnel from a TUN device to a server.
 type OutlineTunnel interface {
 	Tunnel
-	// SetUDPEnabled indicates whether the tunnel and/or the network support UDP traffic.
-	SetUDPEnabled(isUDPEnabled bool)
+
+	// NetworkConnectivityChanged should be called by the native platform when there has been a
+	// network connectivity change. `isConnected` indicates whether the network is connected.
+	// Returns whether UDP proxying is supported in the new network.
+	NetworkConnectivityChanged(isConnected bool) bool
 }
 
 type outlinetunnel struct {
 	*tunnel
 	host         string
-	port         uint16
-	isUDPEnabled bool
+	port         int
+	password     string
+	cipher       string
+	isUDPEnabled bool // Whether the tunnel supports proxying UDP.
 }
 
-// NewOutlineTunnel connects a tunnel to a SOCKS5 server and returns a `Tunnel` object.
+// NewOutlineTunnel connects a tunnel to a Shadowsocks proxy server and returns an `OutlineTunnel`.
 //
-// `host` is the IP or domain of the SOCKS server.
-// `port` is the port of the SOCKS server.
-// `isUDPEnabled` indicates if the SOCKS server and the network support proxying UDP traffic.
+// `host` is the IP or domain of the Shadowsocks proxy.
+// `port` is the port of the Shadowsocks proxy.
+// `password` is the password of the Shadowsocks proxy.
+// `cipher` is the encryption cipher used by the Shadowsocks proxy.
+// `isUDPEnabled` indicates if the Shadowsocks proxy and the network support proxying UDP traffic.
 // `tunWriter` is used to output packets back to the TUN device.
-func NewOutlineTunnel(host string, port uint16, isUDPEnabled bool, tunWriter io.WriteCloser) (OutlineTunnel, error) {
-	if host == "" || tunWriter == nil {
-		return nil, errors.New("Must provide a valid host address, and TUN writer")
+func NewOutlineTunnel(host string, port int, password, cipher string, isUDPEnabled bool, tunWriter io.WriteCloser) (OutlineTunnel, error) {
+	if tunWriter == nil {
+		return nil, errors.New("Must provide a TUN writer")
+	}
+	_, err := shadowsocks.NewClient(host, port, password, cipher)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid Shadowsocks proxy parameters: %v", err.Error())
 	}
 	core.RegisterOutputFn(func(data []byte) (int, error) {
 		return tunWriter.Write(data)
 	})
 	base := &tunnel{tunWriter, core.NewLWIPStack(), true}
-	t := &outlinetunnel{base, host, port, isUDPEnabled}
+	t := &outlinetunnel{base, host, port, password, cipher, isUDPEnabled}
 	t.registerConnectionHandlers()
 	return t, nil
 }
 
-func (t *outlinetunnel) SetUDPEnabled(isUDPEnabled bool) {
-	if t.isUDPEnabled == isUDPEnabled {
-		return
+func (t *outlinetunnel) NetworkConnectivityChanged(isConnected bool) bool {
+	if !isConnected {
+		return false // Don't react to lost network connectivity.
```

Comment:
I thought it could be useful to get notified of lost connectivity (`isConnected == false`) in the future. Can remove it and rename the method if you think it's not adding value.

Reply:
I think right now the function signature is strange.  The name doesn't have anything to do with UDP, but the return value is whether UDP is supported.

If we don't have a specific plan for how to use a notification about lost connectivity, I would lean towards a name like `UpdateUDPSupport()`

Reply:
Got it, done.